modules/network/src/main/java/org/apache/ignite/internal/network/netty/InboundRecoveryHandler.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.network.netty;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import java.util.Collections;
+import org.apache.ignite.internal.network.NetworkMessagesFactory;
+import org.apache.ignite.internal.network.recovery.RecoveryDescriptor;
+import org.apache.ignite.internal.network.recovery.message.AcknowledgementMessage;
+import org.apache.ignite.network.NetworkMessage;
+import org.apache.ignite.network.OutNetworkObject;
+
+/**
+ * Inbound handler that handles incoming acknowledgement messages and sends acknowledgement messages for other messages.
+ */
+public class InboundRecoveryHandler extends ChannelInboundHandlerAdapter {
+    /** Handler name. */
+    public static final String NAME = "inbound-recovery-handler";
+
+    /** Recovery descriptor. */
+    private final RecoveryDescriptor descriptor;
+
+    /** Messages factory. */
+    private final NetworkMessagesFactory factory;
+
+    /**
+     * Constructor.
+     *
+     * @param descriptor Recovery descriptor.
+     * @param factory Message factory.
+     */
+    public InboundRecoveryHandler(RecoveryDescriptor descriptor, NetworkMessagesFactory factory) {
+        this.descriptor = descriptor;
+        this.factory = factory;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        NetworkMessage message = (NetworkMessage) msg;
+
+        if (message instanceof AcknowledgementMessage) {
+            AcknowledgementMessage ackMessage = (AcknowledgementMessage) msg;
+            long receivedMessages = ackMessage.receivedMessages();
+
+            descriptor.acknowledge(receivedMessages);
+        } else if (message.needAck()) {
+            AcknowledgementMessage ackMsg = factory.acknowledgementMessage()
```

Comment:
So, you're saying that every message that needs an ack will receive it individually, right?

Reply:
I believe we should revisit it, maybe not here. Node should passively send its counter every N messages or when idling. I would be cool to compare current approach with a "new" one, by measuring throughput and latency percentiles.

Reply:
This is a quick&dirty approach, the good one will be addressed in this ticket: https://issues.apache.org/jira/browse/IGNITE-16954

Reply:
Ticket is poorly described. Can you elaborate on what you mean by all of it?
Also, I don't see mentioning of batched acknowledgements there

Reply:
Updated the ticket